fdbserver/DataDistribution.actor.cpp

```diff
@@ -1512,219 +1631,690 @@ ACTOR Future<Void> ddGetMetrics(GetDataDistributorMetricsRequest req,
+                                        std::shared_ptr<DDAudit> audit,
+                                        KeyRange range) {
+	TraceEvent(SevInfo, "DDScheduleAuditOnRangeBegin", self->ddId)
+	    .detail("AuditID", audit->coreState.id)
+	    .detail("Range", range)
+	    .detail("AuditType", audit->coreState.getType());
+	state Key begin = range.begin;
+	state KeyRange currentRange;
+	state int64_t issueDoAuditCount = 0;
 
 	while (begin < range.end) {
 		currentRange = KeyRangeRef(begin, range.end);
 		try {
+			TraceEvent(SevInfo, "DDScheduleAuditOnCurrentRange", self->ddId)
+			    .detail("AuditID", audit->coreState.id)
+			    .detail("CurrentRange", currentRange)
+			    .detail("AuditType", audit->coreState.getType());
 			state std::vector<IDDTxnProcessor::DDRangeLocations> rangeLocations =
 			    wait(self->txnProcessor->getSourceServerInterfacesForRange(currentRange));
 
 			state int i = 0;
 			for (i = 0; i < rangeLocations.size(); ++i) {
-				AuditStorageRequest req(audit->id, rangeLocations[i].range, audit->type);
-				if (audit->type == AuditType::ValidateHA && rangeLocations[i].servers.size() >= 2) {
+				AuditStorageRequest req(audit->coreState.id, rangeLocations[i].range, audit->coreState.getType());
+				StorageServerInterface targetServer;
+				// Set req.targetServers and targetServer, which will be
+				// used to doAuditOnStorageServer
+				// Different audit types have different settings
+				if (audit->coreState.getType() == AuditType::ValidateHA) {
+					if (rangeLocations[i].servers.size() < 2) {
+						TraceEvent(SevInfo, "DDScheduleAuditOnRangeEnd", self->ddId)
+						    .detail("Reason", "Single replica, ignore")
+						    .detail("AuditID", audit->coreState.id)
+						    .detail("Range", range)
+						    .detail("AuditType", audit->coreState.getType());
+						return Void();
+					}
+					// pick a server from primary DC
 					auto it = rangeLocations[i].servers.begin();
 					const int idx = deterministicRandom()->randomInt(0, it->second.size());
-					StorageServerInterface& targetServer = it->second[idx];
+					targetServer = it->second[idx];
 					++it;
+					// pick a server from each remote DC
 					for (; it != rangeLocations[i].servers.end(); ++it) {
 						const int idx = deterministicRandom()->randomInt(0, it->second.size());
 						req.targetServers.push_back(it->second[idx].id());
 					}
-					audit->actors.add(doAuditOnStorageServer(self, audit, targetServer, req));
+				} else if (audit->coreState.getType() == AuditType::ValidateReplica) {
+					auto it = rangeLocations[i].servers.begin(); // always compare primary DC
+					if (it->second.size() == 1) {
+						TraceEvent(SevInfo, "DDScheduleAuditOnRangeEnd", self->ddId)
+						    .detail("Reason", "Single replica, ignore")
+						    .detail("AuditID", audit->coreState.id)
+						    .detail("Range", range)
+						    .detail("AuditType", audit->coreState.getType());
+						return Void();
+					}
+					ASSERT(it->second.size() >= 2);
```

Comment:
There could be `single` config?

Reply:
Not sure. I think it would be great to clearly define the boundary of auditStorage and validate the input. As a result, in future, if there is some task running on single config, the audit still be able to work fine.